Add example using MV3 userScripts API

Rob--W opened the pull request:
<!-- 🙌 Thanks for contributing! Adding details below will help us to merge your PR faster. -->

### Description

Add example using MV3 userScripts API. And many other aspects of MV3 development, useful for https://github.com/mdn/webextensions-examples/issues/496.

### Motivation

<!-- ❓ Why are you making these changes and how do they help? -->
This example is part of the documentation of the userScripts API.

### Additional details

<!-- 🔗 Link to documentation, bug trackers, source control, or other places providing more context -->
- https://github.com/w3c/webextensions/issues/279 (History and links to API proposals)
- https://bugzilla.mozilla.org/show_bug.cgi?id=1875475 ("[meta] Implement userScripts API in MV3")
- https://bugzilla.mozilla.org/show_bug.cgi?id=1917000 ("Implement opt-in to userScripts API")

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->

Rob--W commented on userScripts-mv3/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "browser_specific_settings": {
+    "gecko": {
+      "id": "user-script-manager-example@mozilla.org",
+      "strict_min_version": "134.0a1"
```
Note: I put `strict_min_version` 134 because this is the first version of Firefox where the API landed, behind the `extensions.userScripts.mv3.enabled` preference. I may update it to 135 or 136 once we ship it by default on release.

erosman commented:
Note: `"host_permissions"` should include `"file:///*"`, as many users (myself included) run personal userScripts on local files.

Rob--W commented:
> Note: `"host_permissions"` should include `"file:///*"`, as many users (myself included) run personal userScripts on local files.

This is a demo extension, not a full-fledged user script manager. For it to be a complete user script manager, it would have to define the full `GM.*` / `GM_*` APIs, and also have a better UI. That goes beyond demonstrating the extension APIs and is left as an exercise to the reader.

erosman commented:
I understand.. the intension was to demonstrate that the MV3 userScripts API supports "file:///*" as well.

dotproto commented on userScripts-mv3/manifest.json:
```diff
@@ -0,0 +1,21 @@
+  "version": "0.1",
+  "host_permissions": ["*://*/"],
+  "permissions": ["storage", "unlimitedStorage"],
+  "optional_permissions": ["userScripts"],
```
When attempting to load this extension in Nightly I receive the following error:

> Uncaught (in promise) Error: Cannot request permission userScripts since it was not declared in optional_permissions

Channel: Nightly
Version: 135.0a1 (2024-12-30) (aarch64)
OS: macOS 15.1 (24B83)

dotproto replied:
Oops, I missed [Rob's comment below](https://github.com/mdn/webextensions-examples/pull/576/files#r1890374839) indicating that you must currently enable `extensions.userScripts.mv3.enabled` in `about:config`.

dotproto reviewed (changes requested):
In this review I mostly identified small tweaks. The core functionality looks quite solid. Nice work, @Rob--W!

erosman commented on userScripts-mv3/background.js:
```diff
@@ -0,0 +1,176 @@
+
+  // 2. Reset / configure worlds.
+  if (scripts.some(s => s.worldId)) {
+    // When a userscripts need privileged functionality, we run them in a
```
Shouldn’t that be "When a userscript needs" or "When userscripts need"?

Rob--W commented on userScripts-mv3/README.md:
```diff
@@ -0,0 +1,116 @@
+The extension is an example of a
+[user script manager](https://en.wikipedia.org/wiki/Userscript_manager).
+
+This covers the following aspects to extension development:
```
I accepted Simeon's suggestion below instead, which adds "demo" to get to `This demo covers the following aspects to extension development:`

Is this wording OK or is it preferable to use "illustrate"?

rebloor replied:
@Rob--W sorry for the delay two things probably wouldn't use the word demo and "the following" is wordy. "This example demonstrates these aspects of extension development."

Rob--W commented on userScripts-mv3/README.md:
```diff
@@ -0,0 +1,116 @@
+
+- Showing onboarding UI after installation.
+
+- Designing background scripts that can restart repeatedly with minimal
```
To make it easier to read when the file is not rendered.

Rob--W commented on userScripts-mv3/background.js:
```diff
@@ -0,0 +1,176 @@
+browser.permissions.onRemoved.addListener(permissions => {
+  if (permissions.permissions.includes("userScripts")) {
+    // Pretend that userScripts was not available, so that if the permission is
+    // restored, that permissions.onAdded will re-initialize.
```
Rephrased to:

```
    // Pretend that userScripts is not available, to enable permissions.onAdded
    // to re-initialize when the permission is restored.
```

Rob--W commented on userScripts-mv3/background.js:
```diff
@@ -0,0 +1,176 @@
+  // 1. Unregister obsolete scripts.
+  if (scriptsToRemove.length) {
+    let worldIds = scriptsToRemove.map(s => s.id);
+    await browser.userScripts.unregister({ worldIds });
```
Thanks for catching this! I've applied the suggestion and renamed `scriptsToRemove` to `scriptIdsToRemove`.

Rob--W reviewed (commented):
Thanks for the reviews and (comment) suggestions (@dotproto , @rebloor and @erosman). I have just applied them, with the changes visible at https://github.com/Rob--W/webextensions-examples/commits/82331d201638e099c597b04dd2d166f156713c97

I'm going to squash all commits into one to make it ready to merge.

Rob--W commented:
I have updated the PR and set `"strict_min_version": "134"` because the functionality is available in Firefox 134 and 135, just behind a preference. I also added a small section at the end of README.md to mention that.

Although the preference was enabled by default in 136, I did not set it to `136.0a` because AMO does not support extensions with `136.0a` (https://github.com/mozilla/addons/issues/15284). The alternative is `136.0`, but that would prevent the example from running in Firefox Nightly.

rebloor commented on userScripts-mv3/README.md:
```diff
@@ -0,0 +1,123 @@
+demonstrates the userScripts API, the permissions API, `optional_permissions`,
+and Manifest Version 3 (MV3).
+
+This demo covers the following aspects to extension development:
```
```suggestion
This example demonstrates these aspects of extension development:
```

dotproto reviewed (changes requested):
I found a few small changes for polish and consistency.

Rob--W commented on userScripts-mv3/README.md:
```diff
@@ -0,0 +1,123 @@
+Monitoring an optional (userScripts) permission, and dynamically registering
+events and scripts based on its availability:
+
+- The `userScripts` permission is optional and can be granted by the user from:
```
Optional-only is a subset of being an optional permission. Optional-only as a concept only makes a difference at install time; past install time the relevant user experience is identical to optional permissions.

It may be worth referring to the concept of optional-only permissions, but I'm not sure if the current section is the right place for that.

Rob--W commented:
@rebloor and @dotproto : I have applied the suggested changes in https://github.com/mdn/webextensions-examples/commit/23f02c5c324895c7ecaf6ac5dcebd59089ba4bf4, and then squashed the changes in one commit. I haven't made any other changes (other than fixing up one link as explained at https://github.com/mdn/webextensions-examples/pull/576#discussion_r1959663596).

dotproto reviewed (approved):
Looks good!

Rob--W commented:
@rebloor I updated the PR, with the following changes (https://github.com/mdn/webextensions-examples/compare/6bea80c581f43fa3fa9cecd11a93684b86bd70aa..567f4153149b1c84d471e65c5a66e1b122ea2143):

- Dropped "Deprecated" from the old userScripts sample description - now it is just "(Manifest Version 2 only)".
- Updated the README of the old userScripts sample, to emphasize that it is MV2 only, and to refer to the new example. I also updated the link, to point to the `userScripts_legacy/register` documentation on MDN instead of `userScripts/Register` (which will be the home of the MV3 API).
